src/backend/joanie/core/api/client.py

```diff
@@ -358,38 +381,37 @@ def create(self, request, *args, **kwargs):
             )
 
         # Once order has been created, if product is not free, create a payment
-        if product.price > 0:
-            order = serializer.instance
-            payment_backend = get_payment_backend()
-            credit_card_id = serializer.initial_data.get("credit_card_id")
-
-            # if payment in one click
-            if credit_card_id:
-                try:
-                    credit_card = CreditCard.objects.get(
-                        owner=order.owner, id=credit_card_id
-                    )
-                    payment_info = payment_backend.create_one_click_payment(
-                        request=request,
-                        order=order,
-                        billing_address=billing_address,
-                        credit_card_token=credit_card.token,
-                    )
-                except (CreditCard.DoesNotExist, NotImplementedError):
-                    pass
-            else:
-                payment_info = payment_backend.create_payment(
-                    request=request, order=order, billing_address=billing_address
-                )
-
-            # Return the fresh new order with payment_info
+        if product.price > 0 and billing_address:
             return Response(
-                {**serializer.data, "payment_info": payment_info}, status=201
+                {
+                    **serializer.data,
+                    "payment_info": self._create_payment(
+                        serializer.instance, serializer, request, billing_address
+                    ),
+                },
+                status=201,
```

Comment:
The right moment to create the payment is in the transition from submitted to validated?